Edits for WebAssembly 2021 Chapter (#2605)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/webassembly.md

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: WebAssembly

> description: WebAssembly chapter of the 2021 Web Almanac covering Wasm compression, custom sections, most popular instructions, and post-MVP Wasm extensions

Reviewer: @RReverser noticed we didn't have a Meta Description so added this one. You happy with that?

Reviewer 2: ```suggestion
description: WebAssembly chapter of the 2021 Web Almanac covering Wasm compression, section sizes, most popular instructions, and post-MVP proposals
```

maybe a bit better

Let's dive deeper and figure out what those reused libraries are. First, we've tried to deduplicate libraries by content hash alone, but it became quickly apparent that many of those left are still duplicates that differ only by library version.

Then we decided to extract library names from URLs. While it's more problematic in theory due to potential name clashes, it turned out to be a more reliable option for top libraries in practice. We extracted filenames from URLs, removed extensions, minor versions and suffixes that looked like content hashes, sorted the results by number of repetitions and extracted the top 10 modules for each client. For those left, we did manual lookups to understand which libraries those modules are coming from.

> {# TODO - " minor versions and suffixes that looked like content hashes, " - does this mean they were removed, or something else? not clear #}

Author: @RReverser

Reviewer 2: Yes, all of those are meant to be contextually under "removed [...list of things...]".

Reviewer: Added an extra comma in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2605/commits/5129015399710c63060966d803fd930ece2d19e4. I think it's clear so removed the TODO.

What's also interesting, at the other end of the graph—where we were supposed to see the worst savings—we found regressions of up to 1.4 MB. What happened there? How is it possible that Brotli recompression has made things worse for some modules?

As mentioned above, in this article we've used Brotli with compression level 9, but—and we'll admit, we completely forgot about this until this article - it also has levels 10 and 11. Those levels produce even better results in exchange for a steep performance drop-off, as seen, for example, in <a hreflang="en" href="https://quixdb.github.io/squash-benchmark/#results-table">Squash benchmarks</a>. Such trade-off makes them worse candidates for the common on-the-fly compression, which is why we didn't use them in this article and went for a more moderate level 9. However, website authors can choose to compress their static resources ahead of time or cache the compression results, and save even more bandwidth without sacrificing CPU time. Cases like these show up as regressions in our analysis, meaning resources can be and, in some cases, already were optimized even better than we did in this article.

> {# TODO - is there an article supposed to be linked for "this until this article"? #}

Author: @RReverser

Reviewer 2: No, "this article" refers to _this_ article (the one we're in).

  )
}}

> Unsurprisingly, ~74% of the total binary size comes from the compiled code itself, followed by ~19% for embedded static data. Function types, import/export descriptors and such comprise a negligible part of the total size. However, one section type stands out - it's custom sections, which account for ~6.5% of total size in the mobile dataset.

Reviewer 2: I'd argue "most" is important here ("unsurprusingly, most" vs "unsurprisingly, ~74%" - only the fact that it's most is unsurprising, not the precise percentage).

Author: I think it just sounds odd right now, maybe there’s a grammar tweak needed for it to flow better with the word most. Any ideas?

Reviewer 2: I'm probably the wrong person to ask because to me the original sentence sounds fine 😅 Wil moving % to parentheses like "most (~74%)" help?

Reviewer: What about brackets?:

```suggestion
Unsurprisingly, most (~74%) of the total binary size comes from the compiled code itself, followed by ~19% for embedded static data. Function types, import/export descriptors and such comprise a negligible part of the total size. However, one section type stands out - it's custom sections, which account for ~6.5% of total size in the mobile dataset.
```

Author: That works!

Reviewer: Or even move that further down the sentence:

```suggestion
Unsurprisingly, most of the total binary size (~74%) comes from the compiled code itself, followed by ~19% for embedded static data. Function types, import/export descriptors and such comprise a negligible part of the total size. However, one section type stands out - it's custom sections, which account for ~6.5% of total size in the mobile dataset.
```

Author: I like that too

Reviewer 2: First variant seems to be exactly what I proposed, and the 2nd one looks good too. Unfortunately, looks like I can't accept either now because suggestions are marked as outdated after emdash change :(

Reviewer: Updated to second in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2605/commits/5129015399710c63060966d803fd930ece2d19e4

For example, <a hreflang="en" href="https://github.com/WebAssembly/nontrapping-float-to-int-conversions/blob/master/proposals/nontrapping-float-to-int-conversion/Overview.md">non-trapping float-to-int conversions</a> is very similar in spirit to sign-extension operators - it also provides built-in conversions for numeric types to save some code size - but it became uniformly supported only recently with the release of Safari 15. That's why this feature is not yet enabled by default, and most developers don't want the complexity of building and shipping different versions of their WebAssembly module to different browsers without a very compelling reason. As a result, none of the Wasm modules in the dataset used those conversions.

> Other features with zero detected usages - multi-value, reference types and tail calls—are in a similar situation: they could also benefit most WebAssembly use-cases, but they suffer from incomplete compiler and/or engine support.

Reviewer 2: Same question about changes to dash - I think @tunetheweb changed them from ` - ` to mdashes (`—`) in his round of review, now those edits are changing them back and I'm feeling lost about which one is right 😀

Author: Oh shoot lol, hope I didn’t make it worse

Reviewer: [Style guide](https://developers.google.com/style/dashes) says spaceless em-dashes:

> **Em dashes**
> To indicate a break in the flow of a sentence—or an interruption—use an em dash, also known as a long dash. Don't put a space before or after it.

I hate that and think they should have spaces, but that's the style guide we're following!

Reviewer 2: > I hate that 
YES

Author: How do I uncommit all of them?

Reviewer: Let me do a quite search and replace to change them back. Two seconds.

> > I hate that
> YES

I think spaceless em-dashes are more common in America. And unfortunately it's an Google (so I guess American?) style guide.

Reviewer: Put the evil spaceless em-dashes back in: https://github.com/HTTPArchive/almanac.httparchive.org/pull/2605/commits/c60166f8b730f3788c3091803d32b051a10494d6

Among the remaining, used, features, two that are particularly interesting are SIMD and atomics. Both provide instructions for parallelising and speeding up execution at different levels: <a hreflang="en" href="https://v8.dev/features/simd">SIMD</a> allows to perform math operations on several values at once, and atomics provide a basis for <a hreflang="en" href="https://web.dev/webassembly-threads/">multithreading in Wasm</a>. Those features are not enabled by default, require specific use-cases, and multithreading in particular requires using special APIs in the source code as well as additional configuration to make the website <a hreflang="en" href="https://web.dev/coop-coep/">cross-origin isolated</a> before it can be used on the web. As a result, a relatively low usage level is unsurprising, although we expect them to grow over time.

## Conclusion

While WebAssembly is a relatively new and somewhat niche participant on the web, it's great to see its adoption across a variety of websites and use-cases, from simple libraries to large applications.

> In fact, we could see that it integrates so well into the web ecosystem, that many website owners might not even know they already use WebAssembly - to them it looks like any other 3rd-party JavaScript dependency.

Reviewer 2: Why dash here?
